File: apps/driver-android/src/main/kotlin/lk/mageride/driver/onboarding/DriverPermission.kt
package lk.mageride.driver.onboarding

import android.Manifest
import android.annotation.SuppressLint
import android.content.Context
import android.content.Intent
import android.content.pm.PackageManager
import android.net.Uri
import android.os.Build
import android.os.PowerManager
import android.provider.Settings
import androidx.annotation.StringRes
import androidx.core.content.ContextCompat
import androidx.core.content.getSystemService
import lk.mageride.driver.R

/**
 * The four rows of SCR-DA-007, and what each of them actually asks the OS for.
 *
 * D2' §SCR-DA-007 `[DELTA:PLATFORM]` names the Android set exactly: *"ACCESS_BACKGROUND_LOCATION +
 * FOREGROUND_SERVICE_LOCATION + POST_NOTIFICATIONS + battery-optimization intent"*, plus the
 * overlay row the wireframe draws. They are **not** all the same kind of thing, which is why this
 * is an enum with a [kind] rather than four permission strings:
 *
 * * two are runtime permissions granted by a dialog,
 * * one is a system exemption granted by its own settings screen,
 * * one is a special app access granted by another.
 *
 * @property title The row's label.
 * @property rationale The one line under it — *why*, in the driver's language, before the dialog.
 * @property kind How this one is asked for.
 * @property required Whether the driver can reach the dashboard without it. Only the two location
 *   and notification rows gate anything at all, and even they gate **going online**, not Home:
 *   AL-27 puts nothing between Profile Setup and the dashboard.
 */
internal enum class DriverPermission(
    @param:StringRes val title: Int,
    @param:StringRes val rationale: Int,
    val kind: PermissionKind,
    val required: Boolean,
) {

    /**
     * Foreground location. Asked first and on its own: from Android 11 the background grant is
     * only offered *after* this one exists, and asking for both together silently denies both.
     */
    LOCATION(
        title = R.string.permission_location_title,
        rationale = R.string.permission_location_rationale,
        kind = PermissionKind.RUNTIME,
        required = true,
    ),

    /**
     * `ACCESS_BACKGROUND_LOCATION` — publishing with the app behind the lock screen, which is
     * where a mounted handset spends a ride (D6' §3).
     */
    BACKGROUND_LOCATION(
        title = R.string.permission_background_location_title,
        rationale = R.string.permission_background_location_rationale,
        kind = PermissionKind.RUNTIME,
        required = true,
    ),

    /** `POST_NOTIFICATIONS` — E-01's offers live 15 seconds; an unnotifiable driver is undispatchable. */
    NOTIFICATIONS(
        title = R.string.permission_notifications_title,
        rationale = R.string.permission_notifications_rationale,
        kind = PermissionKind.RUNTIME,
        required = true,
    ),

    /**
     * Battery-optimisation exemption. Doze otherwise stretches the D5' §5.2 cadence into minutes,
     * and a vehicle publishing once a minute is a vehicle the map has lost.
     */
    BATTERY(
        title = R.string.permission_battery_title,
        rationale = R.string.permission_battery_rationale,
        kind = PermissionKind.SETTINGS,
        required = false,
    ),

    /** Display over other apps — the offer takeover while the driver is in another app. */
    OVERLAY(
        title = R.string.permission_overlay_title,
        rationale = R.string.permission_overlay_rationale,
        kind = PermissionKind.SETTINGS,
        required = false,
    ),
}

/** How a [DriverPermission] is asked for. */
internal enum class PermissionKind {

    /** A runtime dialog — `ActivityResultContracts.RequestPermission`. */
    RUNTIME,

    /** A settings screen the app can only open; the system decides, and may not come back. */
    SETTINGS,
}

/**
 * Reads and asks for the SCR-DA-007 set.
 *
 * A class rather than free functions so the screen can be driven by a fake in a unit test: every
 * member below is `Context`-bound Android API whose local-unit-test stub returns a default, and a
 * screen that believed those defaults would report every permission granted.
 */
internal open class DriverPermissions(private val context: Context) {

    /** Whether [permission] is currently held. */
    open fun isGranted(permission: DriverPermission): Boolean = when (permission) {
        DriverPermission.LOCATION -> hasRuntime(Manifest.permission.ACCESS_FINE_LOCATION)
        DriverPermission.BACKGROUND_LOCATION -> hasBackgroundLocation()
        DriverPermission.NOTIFICATIONS -> hasNotifications()
        DriverPermission.BATTERY -> isIgnoringBatteryOptimisations()
        DriverPermission.OVERLAY -> Settings.canDrawOverlays(context)
    }

    /**
     * The manifest permission strings to request for a [PermissionKind.RUNTIME] row, or empty when
     * this API level grants it without asking.
     *
     * `POST_NOTIFICATIONS` did not exist before API 33 and `ACCESS_BACKGROUND_LOCATION` was implied
     * by the foreground grant before API 29 — on the URD NFR-22 floor (API 26) both are already
     * held, and requesting a permission the platform does not know is an immediate denial.
     */
    open fun runtimeRequestsFor(permission: DriverPermission): Array<String> = when (permission) {
        DriverPermission.LOCATION -> arrayOf(
            Manifest.permission.ACCESS_FINE_LOCATION,
            Manifest.permission.ACCESS_COARSE_LOCATION,
        )

        DriverPermission.BACKGROUND_LOCATION ->
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q) {
                arrayOf(Manifest.permission.ACCESS_BACKGROUND_LOCATION)
            } else {
                emptyArray()
            }

        DriverPermission.NOTIFICATIONS ->
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU) {
                arrayOf(Manifest.permission.POST_NOTIFICATIONS)
            } else {
                emptyArray()
            }

        else -> emptyArray()
    }

    /**
     * The settings screen for a [PermissionKind.SETTINGS] row, or for a runtime permission the
     * driver has denied permanently — D2's *"denied → Settings deep-link"*.
     */
    @SuppressLint("BatteryLife")
    open fun settingsIntent(permission: DriverPermission): Intent {
        val app = Uri.fromParts("package", context.packageName, null)
        return when (permission) {
            // Google Play restricts this action to apps whose core function needs the exemption.
            // A driver app that stops publishing GPS in Doze has stopped being a driver app, which
            // is the case D7'/D-30's release checklist makes; the lint is suppressed knowingly.
            DriverPermission.BATTERY ->
                Intent(Settings.ACTION_REQUEST_IGNORE_BATTERY_OPTIMIZATIONS, app)

            DriverPermission.OVERLAY ->
                Intent(Settings.ACTION_MANAGE_OVERLAY_PERMISSION, app)

            // Everything else goes to this app's own settings page rather than a global one: a
            // driver sent to the top of Settings has to find the app themselves.
            else -> Intent(Settings.ACTION_APPLICATION_DETAILS_SETTINGS, app)
        }
    }

    private fun hasRuntime(permission: String): Boolean =
        ContextCompat.checkSelfPermission(context, permission) == PackageManager.PERMISSION_GRANTED

    private fun hasBackgroundLocation(): Boolean = if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q) {
        hasRuntime(Manifest.permission.ACCESS_BACKGROUND_LOCATION)
    } else {
        hasRuntime(Manifest.permission.ACCESS_FINE_LOCATION)
    }

    private fun hasNotifications(): Boolean = if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU) {
        hasRuntime(Manifest.permission.POST_NOTIFICATIONS)
    } else {
        true
    }

    private fun isIgnoringBatteryOptimisations(): Boolean =
        context.getSystemService<PowerManager>()?.isIgnoringBatteryOptimizations(context.packageName) == true
}

File: apps/driver-ios/DriverApp/Onboarding/DriverPermission.swift
import CoreLocation
import Foundation
import UIKit
import UserNotifications

/// The two rows of SCR-DI-007, and what each of them actually asks the OS for.
///
/// **Δ Section C — two rows here, five on Android, and the spec says so.** D2' §SCR-DA/DI-007
/// `[DELTA:PLATFORM]` names the two sets separately: *"Android: ACCESS_BACKGROUND_LOCATION +
/// FOREGROUND_SERVICE_LOCATION + POST_NOTIFICATIONS + battery-optimization intent. **iOS:
/// `requestAlwaysAuthorization` + notification auth**"*, and `driver_ios.html` draws exactly those
/// two rows where `driver_android.html` draws four. The missing three are not omissions:
///
/// * **Foreground and background location are one grant on iOS.** `requestAlwaysAuthorization` is a
///   single authorisation with a single row; Android has to ask twice because from API 30 the
///   background grant is only offered after the foreground one exists.
/// * **There is no battery-optimisation exemption.** iOS has no equivalent of Doze's allow-list and
///   no intent to open one. What buys a driver app its background runtime is the `location`
///   background mode, which is declared in `Info.plist` and needs no permission of its own.
/// * **There is no "display over other apps".** An iOS app cannot draw over another; the offer
///   takeover Android puts in an overlay is a notification here, which is the notifications row.
///
/// @property titleKey The row's label.
/// @property rationaleKey The one line under it — *why*, in the driver's language, before the sheet.
/// @property symbolName The wireframe's row glyph, as an SF Symbol.
enum DriverPermission: String, CaseIterable, Identifiable {

    /// `CLLocationManager.requestAlwaysAuthorization` — publishing with the app behind the lock
    /// screen, which is where a mounted handset spends a ride (D6' §3).
    case locationAlways

    /// Notification authorisation — E-01's offers live 15 seconds; an unnotifiable driver is
    /// undispatchable.
    case notifications

    var id: String { rawValue }

    var titleKey: String {
        switch self {
        case .locationAlways: return "permission_location_always_title"
        case .notifications: return "permission_notifications_title"
        }
    }

    var rationaleKey: String {
        switch self {
        case .locationAlways: return "permission_location_always_rationale"
        case .notifications: return "permission_notifications_rationale"
        }
    }

    var symbolName: String {
        switch self {
        case .locationAlways: return "location.fill"
        case .notifications: return "bell.fill"
        }
    }
}

/// Reads and asks for the SCR-DI-007 set.
///
/// A protocol so the screen can be driven by a fake: every member below is a system API whose
/// answer in a unit test is whatever the test host happens to have been granted, and a screen that
/// believed those answers would report a permission state nobody set.
///
/// `@MainActor` because the implementation is: `CLLocationManager` and its delegate are main-thread
/// APIs, and a main-actor method cannot satisfy a nonisolated protocol requirement.
@MainActor
protocol DriverPermissions: AnyObject {

    /// Re-reads whatever the system currently says. Call on appear and on every return to the
    /// foreground — a Settings trip reports back no other way.
    func refresh() async

    /// Whether [permission] is currently held.
    func isGranted(_ permission: DriverPermission) -> Bool

    /// Whether the system will still show a sheet for [permission], or whether only Settings can
    /// change it now — D2's *"denied → Settings deep-link"*.
    func isDeniedPermanently(_ permission: DriverPermission) -> Bool

    /// Asks for [permission]. Answers whether it is held afterwards.
    func request(_ permission: DriverPermission) async -> Bool

    /// Opens this app's own page in Settings.
    func openSettings()
}

/// ``DriverPermissions`` over CoreLocation and `UNUserNotificationCenter`.
///
/// Its own `CLLocationManager` rather than ``PositionService``'s: this one exists to ask a question
/// on one screen, and the publisher's manager is configured for background navigation and starts
/// updating the moment it is authorised. Two managers is the supported shape — authorisation is a
/// property of the app, not of a manager instance.
@MainActor
final class SystemDriverPermissions: NSObject, DriverPermissions, CLLocationManagerDelegate {

    private let manager = CLLocationManager()
    private let pushTokens: PushTokenProvider
    private var pendingLocation: CheckedContinuation<Bool, Never>?

    /// The last notification authorisation read. `UNUserNotificationCenter` answers asynchronously
    /// and a SwiftUI row needs a value now, so it is cached and refreshed by ``refresh()``.
    private var notificationStatus: UNAuthorizationStatus = .notDetermined

    init(pushTokens: PushTokenProvider) {
        self.pushTokens = pushTokens
        super.init()
        manager.delegate = self
    }

    func refresh() async {
        notificationStatus = await UNUserNotificationCenter.current().notificationSettings()
            .authorizationStatus
    }

    func isGranted(_ permission: DriverPermission) -> Bool {
        switch permission {
        case .locationAlways:
            return manager.authorizationStatus == .authorizedAlways
        case .notifications:
            return notificationStatus == .authorized || notificationStatus == .provisional
        }
    }

    func isDeniedPermanently(_ permission: DriverPermission) -> Bool {
        switch permission {
        case .locationAlways:
            // `.authorizedWhenInUse` counts as answered: iOS shows the "keep using in the
            // background?" prompt on its own schedule and a second `requestAlwaysAuthorization`
            // does nothing, so Settings is the only place the driver can complete the upgrade.
            return manager.authorizationStatus != .notDetermined
                && manager.authorizationStatus != .authorizedAlways
        case .notifications:
            return notificationStatus == .denied
        }
    }

    /// Asks the system, once.
    ///
    /// A permission iOS has already refused shows no sheet at all — the call returns immediately
    /// and nothing happens on screen — which is why the screen sends a permanently-denied row to
    /// Settings instead of calling this.
    func request(_ permission: DriverPermission) async -> Bool {
        switch permission {
        case .locationAlways:
            return await requestAlwaysLocation()
        case .notifications:
            // `.alert`, `.sound` and `.badge` — a fifteen-second offer that arrives silently is an
            // offer the driver loses. Provisional authorisation is deliberately not requested: it
            // delivers quietly to the notification centre, which is the one place a ride offer must
            // not go. Registering with APNs on success is ``PushTokenProvider``'s half.
            let granted = await pushTokens.requestAuthorisation()
            await refresh()
            return granted
        }
    }

    func openSettings() {
        guard let url = URL(string: UIApplication.openSettingsURLString) else { return }
        UIApplication.shared.open(url)
    }

    // MARK: - CoreLocation

    /// `requestAlwaysAuthorization` reports through the delegate, not a completion handler, so the
    /// call is bridged onto one continuation. Only ever one in flight: the row that started it is
    /// disabled while it runs.
    private func requestAlwaysLocation() async -> Bool {
        guard manager.authorizationStatus == .notDetermined
            || manager.authorizationStatus == .authorizedWhenInUse
        else {
            return isGranted(.locationAlways)
        }

        return await withCheckedContinuation { continuation in
            pendingLocation = continuation
            manager.requestAlwaysAuthorization()
        }
    }

    nonisolated func locationManagerDidChangeAuthorization(_ manager: CLLocationManager) {
        Task { @MainActor [weak self] in
            guard let self, let continuation = self.pendingLocation else { return }
            self.pendingLocation = nil
            continuation.resume(returning: self.isGranted(.locationAlways))
        }
    }
}
